docs/src/main/asciidoc/mongodb-panache-guide.adoc

```diff
@@ -381,6 +381,30 @@ public class Person extends PanacheMongoEntity {
 Both `findByNameWithPanacheQLQuery()` and `findByNameWithNativeQuery()` methods will return the same result but query written in PanacheQL
 will use the entity field name: `name`, and native query will use the MongoDB field name: `lastname`.
 
+== Query projection
+
+Query projection can be done with the `project(Class)` method on the `PanacheQuery` object that is returned by the `find()` method.
+You can use it to restrict which fields will be returned by the database, for this you need to create a class (a Pojo) that will only contain the projected fields.
+
+[source,java]
+----
+public class PersonName {
+    private String name;
+
+    public String getName(){
+        return name;
+    }
+
+    public void setName(String name){
+        this.name = name;
+    }
+}
+
+// only 'name' will be returned
+PanacheQuery<PersonName> query = Person.find("status ", Status.Alive).project(PersonName.class);
+PanacheQuery<PersonName> query = Person.find("{'status': ?1}", Status.Alive).project(PersonName.class);
```

Comment:
So my main question on that is 

```
public PersonView {
    public String name;
    public LocalDate birthDate;
}

@MongoEntity(collection="ThePerson")
public class Person extends PanacheMongoEntity {
    public String name;

    // will be persisted as a 'birth' field in MongoDB
    @BsonProperty("birth")
    public LocalDate birthDate;

    public Status status;
}

PanacheQuery<PersonName> query = Person.find("status ", Status.Alive).project(PersonName.class);
```

will it return `birthDate` from the `birth` column or will it crash?

Reply:
With the current implementation, it will ask mongodb to project on the `birthDate` field, as it didn't exist, it will return null (no crash).

We should be explicit on the documentation that the attributse of the projection class needs to reflect the attributes on the MongoDB document, if not `@BsonProperty` should be used.

According to me, it's logical, the same kind of mapping (field, getter/setter, annotation) is done on both the projection class or the entity class.

Reply:
With the current implementation, it will ask mongodb to project on the `birthDate` field, as it didn't exist, it will return null (no crash).

We should be explicit on the documentation that the attributse of the projection class needs to reflect the attributes on the MongoDB document, if not `@BsonProperty` should be used.

According to me, it's logical, the same kind of mapping (field, getter/setter, annotation) is done on both the projection class or the entity class.

Reply:
Ah, ok so hear my thinking. With Panache I define entities on which a mapping is defined (with BsonPorperty and friends). And the API especially PanacheQL operate on the properties. (Not the columns unless you go for native query). 
So I anticipated project to be operating on properties too. It is essentially a short hand for return type property projection. 

In that model we would find the property metadata from entity and do the conversion to column.

Reply:
So you would want to use `birthDate` inside the projection class without adding any `@BsonProperty` and having it mapped to the `birth` column because the `project` operator will lookup mapping customization inside the entity and not inside the projection class ?

Thinking about it, it's more logical as the entity defines the mapping related to the database and the projection class (we could have go to a simple interface with only getters instead of a class) only serve to restrict fields. This is a more clearer separation of concerns so I'm OK with it.

I will update my implementation and the documentation accordingly.

Reply:
+1

Reply:
In fact, reviewing my implementation it was the case :)

Using `@BsonProperty` inside the projection class never worked as the query is mapped to a MongoDB `Document` prior to the construction of the `PanacheQueryImpl` and this also means that property/column mapping from the entity was always used :)

Reply:
@emmanuelbernard @FroMage sadly this is not possible as to return a `PanacheQuery<ProjectionClass>` we need to use the `ProjectionClass` to fetch the results from the MongoDB Collection so it must include `@BsonProperty` to map the class property to the database column if needed.

So if field mapping is used, it needs to be placed in the two classes: the entity and the projection one.

I cannot think of a way to avoid this.
My implementation and the documentation now reflect this.

Reply:
Hum hold one, at runtime we don't know that this PanacheQuery object comes from a EntityA repo or active record entity?

Reply:
No, and we don't care as for both Entity usage (active records or repository) mapping is defined the same way: using `@BsonProperty`.
It's the same for JPA where we use `@Column` to customize the property to column mapping for both active records and repository pattern.

Reply:
you lost me. You last message seems to tell me there is no problem. in JPA you project on property name and don't need to replicate your `@Column` annotations

Reply:
You're right ! (I think I lost myself in projection ...). JPA has a lot of different projection capabilites, one is projection DTO and very close to what we want to achieve here and it didn't need any column mapping ...

So if we want to offer the same, I can see no other solution but to use the Entity class when loading data from the database and using a reflection mapper to map from the Entity class to the Projection class.

I don't know if we have such reflection mapper somewhere in the Quarkus code base ?

Reply:
If we know a class is a projection (with an annotation pointing to the origin class) we can copy the column mapping annotations from the origin class to the projection class during transformation.